token-lending/program/src/processor.rs

```diff
@@ -840,168 +759,117 @@ fn process_borrow(
     }
-    if withdraw_reserve.lending_market != repay_reserve.lending_market {
-        return Err(LendingError::LendingMarketMismatch.into());
-    }
-
-    if repay_reserve_info.key == withdraw_reserve_info.key {
-        return Err(LendingError::DuplicateReserve.into());
-    }
-    if repay_reserve.liquidity.mint_pubkey == withdraw_reserve.liquidity.mint_pubkey {
-        return Err(LendingError::DuplicateReserveMint.into());
-    }
-    if &repay_reserve.liquidity.supply_pubkey != repay_reserve_liquidity_supply_info.key {
-        msg!("Invalid repay reserve liquidity supply account");
-        return Err(LendingError::InvalidAccountInput.into());
-    }
-    if &withdraw_reserve.collateral.supply_pubkey != withdraw_reserve_collateral_supply_info.key {
-        msg!("Invalid withdraw reserve collateral supply account");
+    if &obligation.lending_market != lending_market_info.key {
+        msg!("Obligation lending market does not match the lending market provided");
         return Err(LendingError::InvalidAccountInput.into());
     }
-    if &repay_reserve.liquidity.supply_pubkey == source_liquidity_info.key {
-        msg!("Cannot use repay reserve liquidity supply as source account input");
-        return Err(LendingError::InvalidAccountInput.into());
+    if obligation.last_update.is_stale(clock.slot)? {
+        msg!("Obligation is stale and must be refreshed in the current slot");
+        return Err(LendingError::ObligationStale.into());
     }
-    if &withdraw_reserve.collateral.supply_pubkey == destination_collateral_info.key {
-        msg!("Cannot use withdraw reserve collateral supply as destination account input");
-        return Err(LendingError::InvalidAccountInput.into());
+
+    let (liquidity, liquidity_index) =
+        obligation.find_liquidity_in_borrows(*repay_reserve_info.key)?;
+    if liquidity.borrowed_amount_wads == Decimal::zero() {
+        msg!("Liquidity borrowed amount is zero");
+        return Err(LendingError::ObligationLiquidityEmpty.into());
     }
 
-    // accrue interest and update rates
-    assert_last_update_slot(&repay_reserve, clock.slot)?;
-    obligation.accrue_interest(repay_reserve.cumulative_borrow_rate_wads)?;
+    let CalculateRepayResult {
+        settle_amount,
+        repay_amount,
+    } = repay_reserve.calculate_repay(liquidity_amount, liquidity.borrowed_amount_wads)?;
 
-    let RepayResult {
-        integer_repay_amount,
-        decimal_repay_amount,
-        collateral_withdraw_amount,
-        obligation_token_amount,
-    } = obligation.repay(liquidity_amount, obligation_mint.supply)?;
-    repay_reserve
-        .liquidity
-        .repay(integer_repay_amount, decimal_repay_amount)?;
+    if repay_amount == 0 {
+        msg!("Repay amount is too small to transfer liquidity");
+        return Err(LendingError::RepayTooSmall.into());
+    }
 
+    repay_reserve.liquidity.repay(repay_amount, settle_amount)?;
+    repay_reserve.last_update.mark_stale();
```

Comment:
I considered this quite a bit, but I opted for the safe choice of having only a single place where `deposited_value` and `borrowed_value` can be set, and kicked the can on it until I tested the performance. It doesn't seem like it's actually that expensive to mutate and refresh them a bunch of times in a single transaction as the new end to end test demonstrates.

I'm going to add more deposits, borrows, and repays to that test to explore further. I also just made some new changes to skip obligation interest accrual when the cumulative rate hasn't changed, which should make it more performant in this case.

Reply:
Yeah safe and simple is reasonable but for composability, each additional instruction call incurs more cost. So imagine a protocol builds an abstraction over lending, they would have more overhead and would be restricted in what they could do in a single call.

I think there are some operations which don't require the reserve / obligation to be marked stale because internal state could be manually updated. Fine with this optimization coming in a later PR though